HeterogeneousCore/CUDATest/plugins/TestPortableProducerCUDA.cc

```diff
@@ -0,0 +1,58 @@
+#include "CUDADataFormats/Common/interface/Product.h"
+#include "CUDADataFormats/PortableTestObjects/interface/TestDeviceCollection.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "HeterogeneousCore/CUDACore/interface/ScopedContext.h"
+#include "HeterogeneousCore/CUDAServices/interface/CUDAService.h"
+
+#include "TestAlgo.h"
+
+class TestPortableProducerCUDA : public edm::stream::EDProducer<> {
+public:
+  TestPortableProducerCUDA(edm::ParameterSet const& config)
+      : deviceToken_{produces()}, size_{config.getParameter<int32_t>("size")} {}
+
+  void beginStream(edm::StreamID) override {
+    edm::Service<CUDAService> service;
+    if (not service->enabled()) {
+      throw cms::Exception("Configuration") << "CUDAService is disabled.";
```

Comment:
I'd suggest to leave the `beginStream()` out, because the necessary check on `CUDAService` is part of `ScopedContextProduce` (and the CUDA module API is not expected to be evolved further)

Reply:
I didn't remember where the check is in `cms::cuda::ScopedContextProduce`, so I tracked it down:
  - `cms::cuda::ScopedContextProduce(edm::StreamID)` calls `cms::cuda::ScopedContextGetterBase(edm::StreamID)`
  - `cms::cuda::ScopedContextGetterBase(edm::StreamID)` calls `cms::cuda::ScopedContextBase(edm::StreamID)`
  - `cms::cuda::ScopedContextBase(edm::StreamID)` calls `cms::cuda::chooseDevice(edm::StreamID)`
  - `cms::cuda::chooseDevice(edm::StreamID)` does
    ```c++
    edm::Service<CUDAService> cudaService;
    if (not cudaService->enabled()) { ... }
    ```

On the other hand, `cms::alpakatools::ScopedContextProduce<Queue>(edm::StreamID)` eventually calls `cms::alpakatools::chooseDevice<Platform>(edm::StreamID)`, which is simpler:
```c++
    auto const& devices = cms::alpakatools::devices<TPlatform>();
    return devices[id % devices.size()];
```

where `cms::alpakatools::devices<Platform>()` does a lazy initialisation of all the devices for the given `Platform`.
It is normally initialised by the first calls in the corresponding `ALPAKA_ACCELERATOR_NAMESPACE::AlpakaService` - but if the service is not used, it would be initialised by the first modules that tries to use it.

With the aim of having the CUDA and Alpaka versions as close as possible to ease the porting:
  - would it make sense to add a check in `cms::alpakatools::chooseDevice` to throw if service is not available and active ?
  - alternatively, would it make sense to add a check in `cms::alpakatools::chooseDevice` to throw if there are no devices ?

Do you see any downsides of either of these changes ?

Reply:
> * alternatively, would it make sense to add a check in `cms::alpakatools::chooseDevice` to throw if there are no devices ?

I think at least this check would make sense.

> * would it make sense to add a check in `cms::alpakatools::chooseDevice` to throw if service is not available and active ?

About requiring the Service (for Alpaka) I'm less sure. The only impact (wrt just using `cms::alpakatools::devices<TPlatform>()`) currently is if the Service has been enabled or disabled via the configuration, for EDModules that are either explicitly scheduled (i.e. in Path) or consumed by such module (possibly via a longer chain). Not `SwitchProducer` or anything else that is controlled by `process.options.accelerators`. The question is what should such modules behave on a machine with a GPU when `process.options.accelerators` tells not to use a GPU.

Reply:
OK, I've removed the `beginStream()` from the CUDA modules.

I'll open a new PR with the changes to the Alpaka modules and "framework".